crates/matrix-sdk-crypto/src/secret_storage.rs

```diff
@@ -0,0 +1,787 @@
+    MacLength(usize, usize),
+    /// The IV of the secret storage key for the MAC check has an incorrect
+    /// length.
+    #[error("The IV of for the secret storage key MAC check has an incorrect length, expected : {0}, got: {1}")]
+    IvLength(usize, usize),
+    /// The secret storage key is using a unsupported secret encryption
+    /// algorithm. Currently only the [`m.secret_storage.v1.aes-hmac-sha2`]
+    /// algorithm is supported.
+    ///
+    /// [`m.secret_storage.v1.aes-hmac-sha2`]: https://spec.matrix.org/v1.8/client-server-api/#msecret_storagev1aes-hmac-sha2
+    #[error("The secret storage key is using a unsupported secret encryption algorithm")]
+    UnsupportedAlgorithm,
+    /// The passphrase-based secret storage key has an excessively high KDF
+    /// iteration count.
+    #[error(
+        "The passphrase-based secret storage key has an excessively high KDF iteration count: {0}."
+    )]
+    KdfIterationCount(UInt),
+}
+
+/// A secret storage key which can be used to store encrypted data in the user's
+/// account data as defined in the [spec].
+///
+/// The secret storage key can be initialized from a passphrase or from a Base58
+/// encoded string.
+///
+/// To bootstrap a new [`SecretStorageKey`] use the [`SecretStorageKey::new()`]
+/// or [`SecretStorageKey::new_from_passphrase()`] method.
+///
+/// After a new [`SecretStorageKey`] has been created, info about the key needs
+/// to be uploaded to the homeserver as a global account data event. The event,
+/// and event type, for this can be retrieved using the
+/// [`SecretStorageKey::event_content()`] and [`SecretStorageKey::event_type()`]
+/// methods respectively.
+///
+/// # Examples
+/// ```no_run
+/// use matrix_sdk_crypto::secret_storage::SecretStorageKey;
+///
+/// // Create a new secret storage key.
+/// let key =
+///     SecretStorageKey::new_from_passphrase("It's a secret to everybody");
+/// // Retrieve the content
+/// let content = key.event_content();
+/// // Now upload the content to the server and mark the new key as the default one.
+///
+/// // If we want to restore the secret key, we'll need to retrieve the previously uploaded global
+/// // account data event.
+/// let restored_key = SecretStorageKey::from_account_data(
+///     "It's a secret to everybody",
+///     content.to_owned()
+/// );
+/// ```
+///
+/// [spec]: https://spec.matrix.org/v1.8/client-server-api/#secret-storage
+#[derive(Zeroize, ZeroizeOnDrop)]
+pub struct SecretStorageKey {
+    /// Information about the secret storage key.
+    ///
+    /// This is uploaded to the homeserver in a global account data event.
+    #[zeroize(skip)]
+    storage_key_info: SecretStorageKeyEventContent,
+    /// The private key material.
+    secret_key: Box<[u8; 32]>,
+}
+
+impl std::fmt::Debug for SecretStorageKey {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
```

Comment:
you can `use std::fmt` and then simplify this code to:

```suggestion
impl fmt::Debug for SecretStorageKey {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
```

Just a personal taste.